beacon_node/network/src/subnet_service/mod.rs

```diff
@@ -17,12 +32,640 @@ pub enum SubnetServiceMessage {
+            epoch_duration_secs.saturating_mul(spec.epochs_per_sync_committee_period.as_u64()),
+        );
+
+        let track_validators = !config.import_all_attestations;
+        let aggregate_validators_on_subnet =
+            track_validators.then(|| HashSetDelay::new(slot_duration));
+
+        let mut events = VecDeque::with_capacity(10);
+
+        // Queue discovery queries for the permanent attestation subnets
+        if !config.disable_discovery {
+            events.push_back(SubnetServiceMessage::DiscoverPeers(
+                permanent_attestation_subscriptions
+                    .iter()
+                    .cloned()
+                    .map(|subnet| SubnetDiscovery {
+                        subnet,
+                        min_ttl: None,
+                    })
+                    .collect(),
+            ));
+        }
+
+        // Pre-populate the events with permanent subscriptions
+        for subnet in permanent_attestation_subscriptions.iter() {
+            events.push_back(SubnetServiceMessage::Subscribe(*subnet));
+            events.push_back(SubnetServiceMessage::EnrAdd(*subnet));
+        }
+
+        SubnetService {
+            events,
+            beacon_chain,
+            subscriptions: HashSetDelay::new(default_sync_committee_duration),
+            permanent_attestation_subscriptions,
+            scheduled_subscriptions: HashSetDelay::default(),
+            aggregate_validators_on_subnet,
+            waker: None,
+            discovery_disabled: config.disable_discovery,
+            subscribe_all_subnets: config.subscribe_all_subnets,
+            proposer_only: config.proposer_only,
+            log,
+        }
+    }
+
+    /// Return count of all currently subscribed subnets (long-lived **and** short-lived).
+    #[cfg(test)]
+    pub fn subscriptions(&self) -> impl Iterator<Item = &Subnet> {
+        self.subscriptions.iter()
+    }
+
+    #[cfg(test)]
+    pub fn permanent_subscriptions(&self) -> impl Iterator<Item = &Subnet> {
+        self.permanent_attestation_subscriptions.iter()
+    }
+
+    /// Returns whether we are subscribed to a subnet for testing purposes.
+    #[cfg(test)]
+    pub(crate) fn is_subscribed(&self, subnet: &Subnet) -> bool {
+        self.subscriptions.contains_key(subnet)
+    }
+
+    /// Processes a list of validator subscriptions.
+    ///
+    /// This is fundamentally called form the HTTP API when a validator requests duties from us
+    /// This will:
+    /// - Register new validators as being known.
+    /// - Search for peers for required subnets.
+    /// - Request subscriptions for subnets on specific slots when required.
+    /// - Build the timeouts for each of these events.
+    ///
+    /// This returns a result simply for the ergonomics of using ?. The result can be
+    /// safely dropped.
+    pub fn validator_subscriptions(
+        &mut self,
+        subscriptions: impl Iterator<Item = Subscription>,
+    ) -> Result<(), String> {
```

Comment:
I don't see any Error returned from this function, maybe this doesn't need to return a `Result?`